various enhancements

zhubonan opened the pull request:
I think it would be a good time to raise and PR, as discussed in the meeting I have added a few commands and I will add more/change the existing accordingly. 

This PR is a work-in-progress and mainly act as a potential place of discussion and progress tracking.

Changes: 

- [x] Access to `verdi` commands -  I have made the entire `verdi` command accessible in the node shell using the native `click` interface. It is very fast, but there is no autocompletion. I also added expansion so `{}` expands to the current node's pk and `{-1}` expands to that of the last node in the history.
- [x] Access to `verdi shell` using `vshell` command. This start an embedded IPython shell with pre-loaded aiida objects in local scope similar to that of `verdi shell`
- [x] `cd` and `pwd` command for switching directories
- [x] `do_in` , `do_out` stable order and rename `-l/--follow-link-id` and coloring
- [x] make `nodehist_show` display relationships
- [x] add `group_list` and `group_belong` commands
- [x] `-set` options for `label` and `description` commands
- [x] `reload` command for reloading the node and `needs_new_node` decorator so some commands can automatically reload unsealed `ProcessNode`. 
- [x] `comment_add` and `comment_show` commands. 

Konwn Issues:
- `verdi status` will surge CPU usage is rabbitmq is not running - this is probably a problem in the actual `verdi status` command since it was assumed that the python process should terminate after the command execution. 

This PR address: 
- #1 
- #2 
- #3

zhubonan committed:
Added history tracking

zhubonan committed:
Added report command

zhubonan committed:
Added option for in/out to follow links

zhubonan committed:
Polish commands and add readme

zhubonan committed:
Added example aiidashellrc

zhubonan committed:
Added `verdi` command

zhubonan committed:
Added `verdi` command

Now the full power of the verdi command line tool is exposed in
the node shell. Command execution is lightening fast with the
compromise of not having any tab completion. Users should make
their own aliases for various verdi commands.

zhubonan committed:
Renamed commands and added sorting in `in` and `out`

zhubonan committed:
Added `needs_new_node` decorator

zhubonan committed:
Added group list and group belong commands

zhubonan committed:
Added display of node history with links

zhubonan committed:
Stick with _ convention

zhubonan committed:
More informative profile error

zhubonan committed:
Refector the output status and README.md update

zhubonan committed:
Added contributor line

zhubonan committed:
Tweaked output isolation

zhubonan committed:
Added graph generate command

Also updated the decorators indicating the command needs node

zhubonan committed:
Fix Node identifiy compare

zhubonan committed:
Added cd and pwd commands

zhubonan committed:
Enable completion for comment_

zhubonan committed:
Added `verdi_shell` comand

zhubonan committed:
Adjust doc formating

zhubonan commented:
@giovannipizzi @bosonie I have made some major improvement based on the meeting notes. Could you please have the new version in this PR a try and give me some feedback?

Obviously there is no automated testing at the moment... So feedback would be very welcome. 

A major improvement that I think people will like is that the entire `verdi` command line suite can be used with much better responsiveness than that from the shell. In addition, we have command substitution using the currently loaded node or reference to history. So you can use the comand like this:
```
verdi calcjob outputcat {-1} | less
```
without the need to remember the pks.

There is no tab completion though, and it is unlikely to have an easy solution, but people are free to add aliases. For example, I use `vpl` for `verdi process list`. 

We can close #2 and #3 once this is merged.

zhubonan committed:
Added option to cat repo files in byte mode

Useful for zipped files.

bosonie commented:
Hi @zhubonan. This looks great! Seriously!
Only one minor comments. The graphics of `nodehist_show` doesn't satisfy me fully. At the moment I see:
       
        (Dict<21668>@first) nodehist_show
        WorkChainNode<21662> Li 
        🢃  ---  [iteration_01] CALL_CALC
        CalcJobNode<21665>  
        🢃  ---  [output_parameters] CREATE
        Dict<21668>  <-- We are here
        (Dict<21668>@first)

  it would be nice to have an indentation or a bar that helps to distinguish between the output of the `nodehist_show` command and the shell prompt. Something like:

        (Dict<21668>@first) nodehist_show
          |  WorkChainNode<21662> Li 
          |  🢃  ---  [iteration_01] CALL_CALC
          |  CalcJobNode<21665>  
          |  🢃  ---  [output_parameters] CREATE
          |  Dict<21668>  <-- We are here
        (Dict<21668>@first)

I'm not very familiar with `groups` therefore I didn't test this part extensively... the rest seems perfect to me. Thanks!

zhubonan commented:
@bosonie Thanks, I agree it looks much better with the indentation! I will add it soon.

zhubonan committed:
Add indentation for `nodehist_show`

giovannipizzi committed:
Added completion for verdi commands

This calls through the click machinery emulating a BASH terminal,
but it does not call an external subprocess so it's very fast!

I spent quite some effort to make sure it also works with weird
characters like spaces and things to escape. It should be ok now,
but there might be corner cases I didn't check. An easy way to test
is to create a computer with, e.g., a space in the name, and then
try `verdi computer show <TAB>`.
(This also fixes a partially incorrect behavior when passing parameters
to verdi, that was simply splitting on spaces).

I also fixed a raised exception when completing `{}` with missing
nodes in the history, added a newline when quitting using CTRL+D,
and added persistent history with a file that is located in the
folder of the profile (so it's different for different virtual
environments).

giovannipizzi commented:
Indeed, this really looks great! I will now give some feedback on the PR.

> There is no tab completion though, and it is unlikely to have an easy solution, but people are free to add aliases. For example, I use vpl for verdi process list.
I just created a PR on your PR branch (I'm trying to create circular dependencies in GitHub ;-) ) that introduces fast bash completion for verdi!
With this, the tool is perfect, I would say :-D

Actually, I would maybe thing to rename it to `verdi powershell`? 
In principle even if you don't care about the rest of the commands, now you can just call `verdi powershell` and then continue working with verdi commands, that now become super fast?

giovannipizzi commented on .aiidashellrc:
```diff
@@ -0,0 +1,15 @@
+alias create bk backward
```
I would be tempted to remove as much as possible custom aliases, and let instead the user define them, unless we are 100% confident that eveyone wants them.
We could autoload the rc file by expecting to find it in the AiiDA profile directory, similar to what I did in my PR for the history file? I think that is the right place to put it?

(Edit: I saw below this is just intended as an example; then it's OK - but I would still think to just a few defaults that we are kind of convinced we want almost everybody to use)

giovannipizzi commented on README.md:
```diff
@@ -13,7 +13,7 @@ and their attributes, extras and files in the repository.
 
 - nodes are pre-loaded and multiple commands act by default on the "current node"
-- tab completion if *very* fast (ms vs a few hundresd of ms in `verdi`, since the database environment
+- tab completion is *very* fast (ms vs a few hundresd of ms in `verdi`, since the database environment
```
There is another typo I left in `hundresd`, could you fix that as well?

giovannipizzi commented on README.md:
```diff
@@ -80,17 +80,17 @@ The shell by default has no preloaded node, and the shell shows just the current
 
 - Similarly, you can inspect outgoing nodes with `out`. Both for `in` and `out`, you can add a `-t` option to limit results to only one link type.
-
+- Note that the links are labelld by numbers. If you pass `-f <link id>` argument to the `in` command, it will bring you to the node connected by the link.
```
typo (`labelld`). Also, isn't it now `-l`?

giovannipizzi commented on README.md:
```diff
@@ -119,6 +119,8 @@ The shell by default has no preloaded node, and the shell shows just the current
   You can even redirect to file on your computer using `repo_cat aiida.in > test.txt` (and you can then show the file with bash commands by prepending with `!`, see also below): `!ls -l test.txt`, or `!cat test.txt`.
 
+- The report command prints the *report* for a process. It can be used for quick inspection of complex workchains.
+- The `show_hist` command shows a list the visited nodes, and you can quickly go back and forth usnig the `backward` and `forward` commands.  **NOTE** These commands may be renamed later.
```
Names have changed

giovannipizzi commented on README.md:
```diff
@@ -134,6 +136,19 @@ The shell by default has no preloaded node, and the shell shows just the current
+  to quick interaction with `verdi`.
+  An example `.aiidashellrc` file is provided in the repository.
+
+## Konwn issues
```
Typo

giovannipizzi commented on README.md:
```diff
@@ -134,6 +136,19 @@ The shell by default has no preloaded node, and the shell shows just the current
+
+## Konwn issues
+
+- The use of `verdi status` in the node shell without rabbitmq server running will cause CPU usage surge. 
```
Interesting to know. Maybe let's open an issue on the repo once this is merged.

giovannipizzi commented on node_shell.py:
```diff
@@ -10,74 +10,241 @@
 More details in the README.md file.
 
 Author: Giovanni Pizzi, EPFL
+Contributor: Bonan Zhu
```
I guess you did enough work to be considered an author now :-)

giovannipizzi commented on node_shell.py:
```diff
@@ -10,74 +10,241 @@
+from traceback import print_exc
+from pprint import pformat
+from collections import namedtuple
+import contextlib
```
Maybe this is a good point to add pre-commits. We can do it after this PR is merged, for simplicity (maybe you could open an issue on the repo, and I can then do it), in particular regarding sorting of imports.

giovannipizzi commented on node_shell.py:
```diff
@@ -10,74 +10,241 @@
     return wrapper
 
 
+def needs_new_node(f):
```
Maybe we want to put a `always=True` to always reload the node, for those properties that are mutable? Label, description, mtime, comments, extras.
And then put this extra, optional flag in the decorators for these properties?

giovannipizzi commented on node_shell.py:
```diff
@@ -387,10 +851,271 @@ def do_exit(self, arg):  # pylint: disable=unused-argument
+
+    @needs_new_node
+    @cmd2.with_argparser(graph_gen_parser)
+    def do_graph_generate(self, arg):
```
Considering that we now have direct piping to verdi, maybe we should avoid reduplicating commands and trying to replicate the same interface? This will be hard to maintain in the future. And now (also with my PR) there is tab completion + you can use `{}` for the PK. I would thus be tempted to remove this, unless this can give additional interactiveness w.r.t. simply verdi

giovannipizzi commented on node_shell.py:
```diff
@@ -387,10 +851,271 @@ def do_exit(self, arg):  # pylint: disable=unused-argument
+        help='Show also the number of nodes in the group')
+
+    @cmd2.with_argparser(group_parser)
+    def do_group_list(self, args):
```
Same comment here - let's just remove this wrapper and people can now do `verdi group list`

giovannipizzi commented on node_shell.py:
```diff
@@ -387,10 +851,271 @@ def do_exit(self, arg):  # pylint: disable=unused-argument
+
+    @cmd2.with_argparser(group_parser)
+    @needs_node
+    def do_group_belong(self, args):
```
Here maybe I would just call the internal AiiDA command (or directly run a query) without providing all these options? I.e. just provide the list of groups this node is in, simplifying the number of options that you can use?

giovannipizzi commented on node_shell.py:
```diff
@@ -387,10 +851,271 @@ def do_exit(self, arg):  # pylint: disable=unused-argument
+        self.poutput(os.getcwd())
+
+    @with_default_argparse
+    def do_vshell(self, _):
```
Let's call it `verdi_shell` and then people can do an alias?

Also, we should try to clarify the advantage of running `verdi_shell` vs. `verdi shell` as they look a bit different. I guess it's the fact that one gets the current node, right? Also, maybe one can always execute the `%colors` magic, or there are issues with that?

giovannipizzi reviewed (commented):
A few comments

giovannipizzi commented:
In general it looks great! I didn't try out the comments interface but we can improve later in case it's needed.

Instead, I have a final comment on the groups.
- `group_belong` is useful, with the comment I left in
- `group_list` is probably not useful now that we have `verdi` working and with TAB completion

What I had in mind was an interface mirroring the new `verdi group path` (pinging @chrisjsewell for feedback).
The key idea is that if you have the following groups `a`, `a/b`, `a/c`, `a/c/d`, `e`, they behave as folders, so if you do `verdi group path ls` you see only `a` and `e`, if you do `verdi group path ls a` you see `b` and `c`, and so on.

Maybe we could do the following (in the following I speak of groups but I actually refer to group paths):
- keep internally, as we do for the nodes, a current 'group path' and maybe even a history of the recently visited group paths. The default at load would be `/`. If we do a history, it could expose `grouphist_*` commands similar to the `nodehist_*` commands/
- define commands `group_cd`, `group_pwd` (and aliases `gcd`, `gpwd`) to change group using relative paths, or starting with `/` to get a top group.
- make sure `gcd` has working TAB completion
- define command `group_ls` (`gls`) command to show subgroups (with TAB completion)
- have finally an option to show the nodes in the current group (`group_show`, `gshow`). @chrisjsewell by the way, is this possible to do in the current `verdi` command line interface? What is the command? I couldn't find it. Also, is the documentation already accessible? It could be useful for @zhubonan and @bosonie to understand the underlying logic.

One important thing to decide is how to deal with group types. Should we limit to only "user types"? Can we define a syntax (do we already have one?) to load a group of a different type, e.g. `group_name:group_type`?

Since this discussion might become long: maybe let's just have a simple `group_belong` in this PR, and we move possible discussions on the `group path` functionality in an issue (@zhubonan if you see that there starts to be discussion, feel free to move to a separate issue).

zhubonan commented:
@giovannipizzi Thanks for the comments. I will change the codes accordingly. 

I think perhaps it is good to leave the group related stuff to the next PR. I am not familiar with the group path concept yet, since I have not upgraded `aiida-core` to v1.2.0 (need to test the plugins with the new changes in v1.2.0 for safety).

zhubonan committed:
Updated the doc string

zhubonan committed:
Make needs_new_node take always option

zhubonan committed:
Clean up redundant commands

Also enabled colors in `verdi_shell` command.

zhubonan committed:
Simplify group_belong command

Also added myself to the list of authors.

zhubonan committed:
Merge pull request #1 from giovannipizzi/add-verdi-completion

Added completion for verdi commands

zhubonan commented:
@giovannipizzi All done. It is ready to be merged now.

chrisjsewell commented:
> pinging @chrisjsewell for feedback

I think what you said makes sense btw. But yes I'd have to have a play around with this feature to understand better, and happy to have a stab at it in a further PR 👍

zhubonan committed:
Fix a bug with the needs_new_node decorator

giovannipizzi commented on node_shell.py:
```diff
@@ -559,7 +586,7 @@ def do_extrakeys(self, arg):  # pylint: disable=unused-argument
 
-    @needs_new_node
+    @needs_new_node(always=True)
     @with_default_argparse
```
For attributes you don't need always=True, as they cannot be changed after sealing, do you agree?

giovannipizzi commented on node_shell.py:
```diff
@@ -591,7 +618,7 @@ def attrs_choices_method(self):
                                 choices_method=attrs_choices_method)
 
-    @needs_new_node
+    @needs_new_node(always=True)
```
Same here

giovannipizzi commented on node_shell.py:
```diff
@@ -603,7 +630,7 @@ def do_attr(self, arg):
                 arg.attribute_key))
 
-    @needs_new_node
+    @needs_new_node(always=True)
```
Same here

giovannipizzi commented on node_shell.py:
```diff
@@ -1056,10 +1116,10 @@ def load_ipy(cmd2_app, py_bridge):
                 'Entering an embedded verdi shell. Type quit or <Ctrl>-d to exit.\n'
                 'Run Python code from external files with: run filename.py\n'
-                'If the color is not on, you can enable it using line magic \'%colors linux\'\n'
                 'The loaded node \'{}\' can be accessed with \'current_node\'\n'
```
This line might be skipped if the current node does not exist

giovannipizzi commented on node_shell.py:
```diff
@@ -1035,6 +1092,9 @@ def load_ipy(cmd2_app, py_bridge):
             from IPython import embed
+            from traitlets.config import get_config
+            conf = get_config()
+            conf.InteractiveShellEmbed.colors = "Linux"
```
I don't know if this is the reason, but now I get white text, and my default shell has white background... Is there a way to just use the same default as ipython would do?

giovannipizzi reviewed (changes requested):
Thanks a lot @zhubonan !
I just have a very few final suggestions

zhubonan committed:
Changed as requested

zhubonan commented:
@giovannipizzi Thanks, all done, please do check if the colors are OK on OSX!

giovannipizzi commented:
Thanks a lot!!!!
